_RELEASE/Packs/cube/Scripts/utils.lua

```diff
@@ -1,21 +1,552 @@
+	]]
+	return (FPS * 60)/bpm;
+end
+
+function BPMtoSPB(bpm)
+	--[[
+	Converts Beats Per Minute (BPM) into period in Seconds Per Beat (SPB)
+	]]
+	return FPS/bpm;
 end
 
--- clamp: clamps a number between two values
 function clamp(input, min_val, max_val)
+	--[[
+	Clamps a number "input" between two values. The value can not go 
+	below min_val and can not go above max_val.
+	]]
 	if input < min_val then
 		input = min_val
 	elseif input > max_val then
 		input = max_val
 	end
 	return input
 end
+
+local function fromHS(h, s)
+	--[[
+	Helper function for "fromHSV()", which translates from Hue and Saturation
+	to RGB values.
+	]]
+	-- Find a color from hue and saturation.
+	h = (h % 360)/60;
+	local i, f, g, t;
+	i, f = math.modf(h)
+	g = 1 - f; -- For descending gradients
+	t = 1 - s; -- Minimum color intensity based on saturation
+	f, g = s * f + t, s * g + t; -- Apply saturation
+	
+		if i == 0 then return 1, f, t;
+	elseif i == 1 then return g, 1, t;
+	elseif i == 2 then return t, 1, f;
+	elseif i == 3 then return t, g, 1;
+	elseif i == 4 then return f, t, 1;
+	elseif i == 5 then return 1, t, g; 
+	end
+	return 1, 1, 1; -- Fallback
+end
+
+function fromHSV(h, s, v)
+	--[[
+	Converts HSV color values to RGB color values in a 0 - 255 range (alpha not included).
+	]]
+	-- Saturation and Value are optional parameters
+	s = s or 1;
+	v = v or 1;
+	
+	local r, g, b = fromHS(h, s);
+	r = math.floor(r * v * 255);
+	g = math.floor(g * v * 255);
+	b = math.floor(b * v * 255);
+	return r, g, b;
+end
+
+function getAbsolutePosition(x, y)
+	--[[
+	Returns coordinates (x, y) adjusted with the level rotation using polar coordinate math.
+	Implementation by Oshisaure.
+	]]
+	local r, a = (x ^ 2 + y ^ 2) ^ 0.5, math.atan(y, x);
+	a = a + math.rad(l_getRotation());
+	return r * math.cos(a), r * math.sin(a);
+end
+
+function getSign(number) 
+	--[[
+	Takes a number and returns it's sign value.
+	
+	Negatives return -1, positives return 1, and zero returns 0.
+	]]
+	if number < 0 then
+		return -1;
+	elseif number > 0 then
+		return 1;
+	end
+	return 0;
+end
+
+-- A data structure that stores results for prime numbers as a way to "cache" results.
+-- This helps make isPrime run faster by knocking out calculations that were already done.
+local prime_memoization = {
+	[0] = false,
+	[1] = false,
+	[2] = true
+};
+function isPrime(integer, divisor)
+	--[[
+	Determines if an integer is a prime number or not.
+	]]
+	-- Makes the divisor parameter optional. By default, it should start at 2.
+	divisor = divisor or 2; 
+
+	if (prime_memoization[integer] ~= nil) then
+    	return prime_memoization[integer];
+  	end
+	
+	if (integer % divisor == 0) then
+    	prime_memoization[integer] = false;
+		return false;
+	elseif (divisor * divisor > integer) then
+    	prime_memoization[integer] = true;
+		return true;
+	end
+	-- Make a recursive call to try other factors.
+	return isPrime(integer, divisor + 1); 
```

Comment:
Does this have to be recursive? Using a for loop instead of the divisor argument would probably be a lot easier on the memory.

Reply:
I probably at the time wrote this in my mathExtended module to show off that recursive functions could be useful, but looking at this code now I can see why doing this more iteratively wouldn't make a difference. It's not like it is dependent on recursiveness.